### DISCUSSION TITLE
Consider adding support for "whispering" in group rooms

### DISCUSSION THREAD

**Discussion by richvdh** [element-hq/element-meta#943] [Ideas]
Apparently XMPP has the ability to send messages to a specific subset of users in a larger room; we may wish to consider supporting this too.

Advantages of the feature are that it can be useful to send a quick/offtopic message to a subset of users in a channel about something small or only affecting them, especially in high-traffic rooms, without necessarily starting a new discussion.

Concerns about such a feature are that it would provide a ready spam/harassment vector: people in large rooms may use it to send inappropriate messages out-of-sight of moderators. We would need to make sure that such a feature came alongside tools for managing such misuse (eg blocking, reporting).

**Comment by richvdh** [comment]
A potential implementation of this might be via per-message ACLs.

**Comment by ara4n** [comment]
Per-Message ACLs are basically https://github.com/matrix-org/matrix-doc/issues/482.  In terms of this particular application of them, it feels of very very questionable value given you can just create a room if you want to talk behind people's backs like this.

**Comment by heyakyra** [comment]
It can be useful to send a quick/offtopic message to a subset of users in a channel about something small or only affecting them, especially in high-traffic rooms, without necessarily starting a new discussion. Basically a nifty little way to transparently but unobtrusively post an aside, interjection, or tangent, which may either not require response or itself be the invitation to join/create a new room. 

What's missing from the expectation of creating a room to talk to someone or responding to invitations, is the ability to trace where the conversation stemmed from and how the room's creator found you. On Riot and elsewhere, I've received DM invitations stemming my presence in group channels that i answer (it's natural to assume good reason for contact) but end up being...unwarranted. While this feature wouldn't solve that issue, it at least reflects some existing social norms and might provide the *possibility* of a convention of expressing a wish to start a private conversation before actually doing so. 

  _**\<sidenote>** I think the only thing that could truly solve that issue for me, is a situation where every connection to a room would create a new offshoot/derivative ID so that ever subsequent contact can be traced back. The "root" identity would not be shareable. This would allow people to share different contact IDs in different places, and be able to prioritize or respond to them differently, or even disable one. The major caveat is that this might require [UUIDs](https://github.com/matrix-org/synapse/issues/1948) instead of usernames. (Edit: as in vector-im/element-web#5336) **\</sidenote>**_

To address concerns of this encouraging bad behavior, making sure that the "whispers" (maybe better to call them something that sounds less intimate like "sidenotes") are not actually private. That means it should discourage being easy to have entire whispered/secret conversations in a channel, but still be easier than creating a new room and inviting users to it. If this can be encouraged (through interface or otherwise) as a way to start talking to someone, bad actors may be more easily dealt with by moderators.

**Comment by Half-Shot** [comment]
>Per-Message ACLs are basically matrix-org/matrix-doc#482. In terms of this particular application of them, it feels of very very questionable value given you can just create a room if you want to talk behind people's backs like this.

This *does* feel more like a UI thing which essentially translates into I want to message someone while still keeping the context of the room. You could probably achieve the same effect by:

* Hooking up a button/command/whatever to enable whisper mode to a user. Probably single use.
* Sending a message through a per-existing room or creating and inviting the user to a new one. 
* Attaching some metadata that the message should be inlined into the whispered to room.

The pros to this are:
* Whispers require the user to accept the room, which is probably desirable for people who don't want to be pestered.
* Whispers do not disrupt the room history, and use existing systems to keep the message private. This also allows you to use E2E whispering even if the room isn't E2E enabled (which is probably the case for 99% of public rooms).
* You can always trace your conversation between yourself and the target, even in a room that is quite loud.

Disadvantages are that you are now injecting events into a view of a room from another room which might cause confusion, especially if the recipient tries to respond to it publicly.

**Comment by hukaze** [comment]
>Per-Message ACLs are basically matrix-org/matrix-doc#482. In terms of this particular application of them, it feels of very very questionable value given you can just create a room if you want to talk behind people's backs like this.

Its not about talking behind its about controlling who can see what.. In the case of a support room that might have hundreds of end users that don't need to know that user x is having a problem with their printer and that the CxO is having a problem with their mobile.. The room could support, for example, 3 mods that can see and respond to all messages with 'quote' to directly respond to the specific user and with non-quoted message to send a message to all users; e.g. maintenance tonight from 8-10pm, etc..

**Comment by mjattiot** [comment]
@ara4n the main use case I see for whispering is when asking a bot for help :
In Slack when typing @my_bot help, the bot is answering with a special message only visible to the sender so it doesn't spam the channel.
We also found a use case when using forms messages. When a user is submitting a form, the bot can be interested in whispering some errors or information to him.

What do you think ?

**Comment by Mikaela** [comment]
Many IRC networks have support for ChanServ ENTRYMSGs or other bots that send a message in private upon joining a channel. I like how WeeChat implements "whispers" through `NOTICE nick [#channel] Welcome to #channel` and shows them in the correct buffers.

I think copying the WeeChat behaviour would be beneficial to matrix-appservice-irc, so people wouldn't be confused on "who is ChanServ and why it wants to message me" or "who is ]o[" etc. I think it might also actually make Matrix users read those messages and decrease some frustation IRC users have with them.

**Comment by bkil** [comment]
I think support of threading and allowing users to ignore specific threads would offer a more straight forward and universal option.

**Comment by vertigo220** [comment]
In addition to examples already given, I can think of at least several more where whispering would be useful, to allow some individuals in a room with many people to quickly say something between them where either it would just add unnecessary clutter to the room conversation or they may simply not want it to be in the public conversation, but easier than having to create a separate room or DM. And there's no need to make them non-private, and they can and probably should be encrypted if possible, even if the room isn't (though this should probably be a setting). If there's abuse/harassment, all it would take is a copy/paste or screenshot shown to an admin to make it known so it can be dealt with.
